snuba/query/joins/pre_processor.py

```diff
@@ -0,0 +1,166 @@
+from __future__ import annotations
+
+from typing import Mapping, MutableMapping, NamedTuple, Set
+
+from snuba.datasets.entities import EntityKey
+from snuba.datasets.entities.factory import get_entity
+from snuba.query.data_source.join import (
+    IndividualNode,
+    JoinClause,
+    JoinNode,
+    JoinVisitor,
+)
+from snuba.query.data_source.simple import Entity
+
+
+class QualifiedCol(NamedTuple):
+    entity: EntityKey
+    column: str
+
+
+# An adjacency set representation of a graph that represents
+# semantic equivalence between columns across entities.
+# Each node is a QualifiedCol instance, which represents entity
+# and column.
+# Each edge represent an equivalence between two nodes.
+EquivalencesGraph = MutableMapping[QualifiedCol, Set[QualifiedCol]]
+
+
+def _entity_from_node(node: IndividualNode[Entity]) -> EntityKey:
```

Comment:
Maybe this should be an `@property` on IndividualNode. It is useful outside of this.